nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/NiFiAtlasClient.java

```diff
@@ -201,12 +202,12 @@ public NiFiFlow fetchNiFiFlow(String rootProcessGroupId, String clusterName) thr
         nifiFlow.setUrl(toStr(attributes.get(ATTR_URL)));
         nifiFlow.setDescription(toStr(attributes.get(ATTR_DESCRIPTION)));
 
-        nifiFlow.getQueues().putAll(toQualifiedNameIds(toAtlasObjectIds(nifiFlowEntity.getAttribute(ATTR_QUEUES))));
-        nifiFlow.getRootInputPortEntities().putAll(toQualifiedNameIds(toAtlasObjectIds(nifiFlowEntity.getAttribute(ATTR_INPUT_PORTS))));
-        nifiFlow.getRootOutputPortEntities().putAll(toQualifiedNameIds(toAtlasObjectIds(nifiFlowEntity.getAttribute(ATTR_OUTPUT_PORTS))));
+        nifiFlow.getQueues().putAll(fetchFlowComponents(TYPE_NIFI_QUEUE, nifiFlowReferredEntities));
+        nifiFlow.getRootInputPortEntities().putAll(fetchFlowComponents(TYPE_NIFI_INPUT_PORT, nifiFlowReferredEntities));
+        nifiFlow.getRootOutputPortEntities().putAll(fetchFlowComponents(TYPE_NIFI_OUTPUT_PORT, nifiFlowReferredEntities));
 
         final Map<String, NiFiFlowPath> flowPaths = nifiFlow.getFlowPaths();
-        final Map<AtlasObjectId, AtlasEntity> flowPathEntities = toQualifiedNameIds(toAtlasObjectIds(attributes.get(ATTR_FLOW_PATHS)));
+        final Map<AtlasObjectId, AtlasEntity> flowPathEntities = fetchFlowComponents(TYPE_NIFI_FLOW_PATH, nifiFlowReferredEntities);
 
         for (AtlasEntity flowPathEntity : flowPathEntities.values()) {
```

Comment:
Feels strange to leave the handling of the flowpaths as it is. We retrieve only the `ACTIVE` ones, but we discard their `referredEntities`. So when it comes to their inputs/ouputs, we fall back to the old logic and retrieve them all and filter out the `DELETED` ones afterwards.

Even more, to me it seems if we actually didn't set `minExtInfo` to `true` but left it on `false`, we could get away with a single REST call, because in that case each of the `referredEntities` would contain `relationshipAttributes` that also have a `guid` and `entityStatus` (`ACTIVE` or `DELETED`) attribute.

Reply:
Retrieving the flow and its components is composed of 3 steps:

1. get the flow
2. get the flow paths (and other subcomponents of the flow)
3. get the inputs / outputs of the flow paths

Input / output data was returned in all 3 steps which was definitely overkill.

[NIFI-6945](https://issues.apache.org/jira/browse/NIFI-6945) was about excluding the input/output data from step 1. There are 2 reasons for doing this: Getting all inputs/outputs of all flow paths in a single query can lead to a huge response json from Atlas (extremely high response time / timeout). Furthermore, there is no way to query only the ACTIVE `referredEntities` from Atlas. So in case of `minExtInfo=true`, all the deleted flow paths were retrieved with all their inputs/outputs unnecessarily (they will be filtered out in step 2). So I believe that having multiple but smaller requests in step 2 is better than having one huge request in step 1.

Your comment about falling back to the old logic (in step 3) is reasonable. Input / output data is still retrieved multiple times in step 2 and 3. My plan is to create a separate ticket for it. I would keep the scope of this ticket on the active/delete filtering of the flow paths only.